sdk/eventhub/Azure.Messaging.EventHubs/src/Authorization/SharedAccessSignature.cs

```diff
@@ -0,0 +1,435 @@
+        /// </summary>
+        ///
+        /// <returns>A new copy of <see cref="SharedAccessSignature" />.</returns>
+        ///
+        internal SharedAccessSignature Clone() =>
+            new SharedAccessSignature()
+            {
+                Resource = this.Resource,
+                SharedAccessKeyName = this.SharedAccessKeyName,
+                SharedAccessKey = this.SharedAccessKey,
+                ExpirationUtc = this.ExpirationUtc,
+                Value = this.Value
+            };
+
+        /// <summary>
+        ///   Builds the shared accesss signature value, which can be used as a token for
+        ///   access to the Event Hubs service.
+        /// </summary>
+        ///
+        /// <param name="audience">The audience scope to which this signature applies.</param>
+        /// <param name="sharedAccessKeyName">The name of the shared access key that the signature should be based on.</param>
+        /// <param name="sharedAccessKey">The value of the shared access key for the signagure.</param>
+        /// <param name="expirationUtc">The date/time, in UTC, that the signature expires.</param>
+        ///
+        /// <returns>The value of the shared access signature.</returns>
+        ///
+        protected virtual string BuildSignature(string   audience,
+                                                string   sharedAccessKeyName,
+                                                string   sharedAccessKey,
+                                                DateTime expirationUtc)
+        {
+            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(sharedAccessKey)))
+            {
+                var encodedAudience = WebUtility.UrlEncode(audience);
+                var expiration = Convert.ToString(ConvertToUnixTime(expirationUtc), CultureInfo.InvariantCulture);
+                var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes($"{ encodedAudience }\n{ expiration }")));
+
+                return String.Format(CultureInfo.InvariantCulture, "{0} {1}={2}&{3}={4}&{5}={6}&{7}={8}",
+                    AuthenticationTypeToken,
+                    SignedResourceToken, encodedAudience,
+                    SignatureToken, WebUtility.UrlEncode(signature),
+                    SignedExpiryToken, WebUtility.UrlEncode(expiration),
+                    SignedKeyNameToken, WebUtility.UrlEncode(sharedAccessKeyName));
+            }
+        }
+
+        /// <summary>
+        ///   Parses a shared access signature into its component parts.
+        /// </summary>
+        ///
+        /// <param name="sharedAccessSignature">The shared access signature to parse.</param>
+        ///
+        /// <returns>The set of composite properties parsed from the signature.</returns>
+        ///
+        protected virtual (string KeyName, string Resource, DateTime ExpirationUtc) ParseSignature(string sharedAccessSignature)
```

Comment:
We usually avoid using tuples in public APIs

Reply:
Why have this method as virtual? Can't tests call `SharedAccessSignature(string sharedAccessSignature, string sharedAccessKey)` ctor and observe resulting objects properties?

Reply:
The API isn't meant to be public.   It's meant to allow consumed by tests only.   Flipping the type to `internal` should resolve, which I'll be doing.    I don't believe that I used the override behavior that I had planned to here either and am likely to kill `virtual` as well.